proof_of_stake/src/lib.rs

```diff
@@ -2708,36 +2710,40 @@ fn make_bond_details(
         .cloned()
         .unwrap_or_default();
     let amount = token::Amount::from_change(change);
-    let slashed_amount =
-        slashes
-            .iter()
-            .fold(None, |acc: Option<token::Amount>, slash| {
-                if slash.epoch >= start {
-                    let validator_slashes =
-                        applied_slashes.entry(validator.clone()).or_default();
-                    if !prev_applied_slashes
-                        .iter()
-                        .any(|s| s.clone() == slash.clone())
-                    {
-                        validator_slashes.push(slash.clone());
-                    }
-                    return Some(
-                        acc.unwrap_or_default()
-                            + mult_change_to_amount(slash.rate, change),
-                    );
-                }
-                acc
-            });
-    let slashed_amount =
-        slashed_amount.map(|slashed| cmp::min(amount, slashed));
+    let mut slash_rates_by_epoch = BTreeMap::<Epoch, Decimal>::new();
+
+    let validator_slashes =
+        applied_slashes.entry(validator.clone()).or_default();
+    for slash in slashes {
+        if slash.epoch >= start {
+            let cur_rate = slash_rates_by_epoch.entry(slash.epoch).or_default();
+            *cur_rate = cmp::min(Decimal::ONE, *cur_rate + slash.rate);
+
+            if !prev_applied_slashes
+                .iter()
+                .any(|s| s.clone() == slash.clone())
+            {
+                validator_slashes.push(slash.clone());
+            }
+        }
+    }
+
+    let amount_after_slashing = token::Amount::from_change(
+        get_slashed_amount(params, amount, &slash_rates_by_epoch).unwrap(),
+    );
+    let slashed_amount = if amount == amount_after_slashing {
```

Comment:
let's instead call `get_slashed_amount` only when `!slash_rates_by_epoch.is_empty()` as it does some alloc that we can avoid most of the time

Reply:
also in 65e375c5c44873b9df48482bff0316bf3fc3b54a